### ISSUE TITLE
sqliteral equiv of tiny_sqlite's unpack()?

### ISSUE THREAD

**Issue by ITwrx** [olliNiinivaara/SQLiteral#6]
I'm trying to test moving from tiny_sqlite to sqliteral, but I need to access column data from row(s) of a table that has more than one column and type, but i'm not seeing how to do that, as I'm not seeing any equivalent of tiny_sqlite's unpack().

[tiny_sqlite's unpack](https://gulpf.github.io/tiny_sqlite/tiny_sqlite.html#unpack%2CResultRow%2Ctypedesc%5BT%5D)

a snippet of my tiny_sqlite model code using unpack():

```
for row in db.iterate("SELECT id, title, link_title, nav_menu_inclusion, landing_type, thumbs_type, parent_id, status FROM SiteSections"):
  let (id, title, link_title, nav_menu_inclusion, landing_type, thumbs_type, parent_id, status) = row.unpack((int, string, string, string, string, string, int, string))
```

Thanks

**Comment by olliNiinivaara** [comment]
For optimal performance, it is best to access field values by column indices using getCString, getInt etc. procs, because they are zero-copy. But please check the _toTuple_  proc in the example below.

```nim

from db_connector/sqlite3 import column_type, SQLITE_INTEGER, SQLITE_FLOAT, SQLITE_TEXT, SQLITE_BLOB
import sqliteral

const Schema = "CREATE TABLE IF NOT EXISTS Example(string TEXT, int INTEGER, float REAL, big BLOB)"
type
  SqlStatements = enum
    Insert = """INSERT INTO Example(string, int, float, big) VALUES(?, ?, ?, ?)"""
    SelectAll = "SELECT * FROM Example"
  RowTuple = tuple[stringfield: DbValue, intfield: DbValue, floatfield: DbValue, bigfield: DbValue]
var db: SQLiteral

proc toTuple*[T: tuple](prepared: PStmt, _: typedesc[T]): T =
  var col = 0.int32
  for value in fields(result):
    let datatype = column_type(prepared, col)
    case datatype
      of SQLITE_INTEGER: value = toDb(getInt(prepared, col))
      of SQLITE_FLOAT: value = toDb(getFloat(prepared, col))
      of SQLITE_TEXT: value = toDb(getCString(prepared, col))
      of SQLITE_BLOB: value = toDb(getSeq(prepared, col))
      else: discard
    col += 1

db.openDatabase("example.db", Schema)
db.prepareStatements(SqlStatements)
db.transaction: db.exec(Insert, "somestring", 15, 3.14, @[byte 65,66,67])
for row in db.rows(SelectAll): echo row.toTuple(RowTuple)
db.close()

```

**Comment by ITwrx** [comment]
Thanks a lot for that comprehensive example. I'll leave this open for now while i try it out.

**Comment by ITwrx** [comment]
I'm trying this out, but while your `toTuple()` example does tuple-ize the DB row, it leaves the fields as `DBValues`. The `unpack()` proc from `tiny_sqlite` also lets you set the new types for the fields, so you can use them in seqs, for instance. So far, i'm still not clear on how to get plain ol' ints and strings from the sqliteral db rows.

**Comment by ITwrx** [comment]
I'm not seeing any equivalent to tiny_sqlite's `fromDbValue()` in sqliteral.

**Comment by olliNiinivaara** [comment]
You access DbValue's value as with any [object variant](https://nim-lang.org/docs/tut2.html#object-oriented-programming-object-variants):

```nim
for row in db.rows(SelectAll):
  let t = row.toTuple(RowTuple)
  var s = t.stringfield.textVal
  var i = t.intfield.intVal
  var f = t.floatfield.floatVal
  var bs = t.bigfield.blobVal
```

Of course you can also convert row result to tuple manually (tiny_sqlite's unpack just automates this, the end result is the same):
```nim
for row in db.rows(SelectAll):
  var t: tuple[stringfield: string, intfield: int, floatfield: float, bigfield: seq[byte]]
  t.stringfield = row.getString(0); t.intfield = row.getInt(1); t.floatfield = row.getFloat(2); t.bigfield = row.getSeq(3)
  echo t
```


But in most cases there is no need to copy the row values to other variables at all, because "the tuple values are already there", accessible with the row.getXXX(col) getters.

**Comment by ITwrx** [comment]
Thanks for the link and the example code! I had recently skipped reading about object variants, thinking it unnecessary at the time. :) I ended up using your last suggested approach:

```Nim
proc getAllSiteSections*(): seq[SiteSection] =
  var siteSections: seq[SiteSection]
  const Schema = "CREATE TABLE IF NOT EXISTS SiteSections(id INTEGER PRIMARY KEY, title TEXT NOT NULL, link_title TEXT, nav_menu_inclusion TEXT, landing_type TEXT, parent_id INTEGER, status TEXT)"
  type
    SqlStatements = enum
      SelectAll = "SELECT * FROM SiteSections"
  var db: SQLiteral
  db.openDatabase("app.db", Schema)
  db.prepareStatements(SqlStatements)
  for row in db.rows(SelectAll): 
    var newSiteSection: SiteSection
    newSiteSection = SiteSection(id: row.getInt(0), title: row.getString(1), linkTitle: row.getString(2), navMenuInclusion: row.getString(3), landingType: row.getString(4), parentId: row.getInt(5), status: row.getString(6))
    siteSections.add(newSiteSection)
  db.close()
  return siteSections 
```
It seems i was supposed to use `prepareStatements(SqlStatements)` even with only one statement, so i'm doing that now.


Now my ~~C~~ R ~~U~~ ~~D~~ test works!

**Comment by ITwrx** [comment]
hmm, i get a 500 with the above db code, but only after the third refresh. sqliteral doesn't give any errors, but debug shows:

```
PRAGMA encoding = 'UTF-8'
PRAGMA foreign_keys = ON
PRAGMA locking_mode = EXCLUSIVE
PRAGMA journal_mode = WAL
CREATE TABLE IF NOT EXISTS SiteSections(id INTEGER PRIMARY KEY, title TEXT NOT NULL, link_title TEXT, nav_menu_inclusion TEXT, landing_type TEXT, parent_id INTEGER, status TEXT)
app.db opened
SELECT * FROM SiteSections
SELECT * FROM SiteSections
app.db closed
```

I wonder why `SELECT * FROM SiteSections` is run twice. 
Please let me know if you would prefer i open a new issue.
Thanks

**Comment by olliNiinivaara** [comment]
The database must be opened only once, when the program starts (and closed only when the program is being shut down). Also, the ```prepareStatements``` proc must be called only once per each thread that is going to access the database. This may a bit tricky to achieve, but the  [the main example](https://olliniinivaara.github.io/SQLiteral/#example-using-json-extensions) tries to explain the way forward. Therefore, the implementation should resemble something like this:

```nim

type
  SqlStatements = enum
    SelectAll = "SELECT * FROM SiteSections"

var
  db*: SQLiteral
  prepared {.threadvar.}: bool

proc startProgram() =
  const Schema = "CREATE TABLE IF NOT EXISTS SiteSections(id INTEGER PRIMARY KEY, title TEXT NOT NULL, link_title TEXT, nav_menu_inclusion TEXT, landing_type TEXT, parent_id INTEGER, status TEXT)"
  db.openDatabase("app.db", Schema)

proc getAllSiteSections*(): seq[SiteSection] =
  {.gcsafe.}:
    if not prepared:
      db.prepareStatements(SqlStatements)
      prepared = true
    for row in db.rows(SelectAll): 
      result.add(SiteSection(id: row.getInt(0), title: row.getString(1), linkTitle: row.getString(2), navMenuInclusion: row.getString(3), landingType: row.getString(4), parentId: row.getInt(5), status: row.getString(6))) 

proc closeProgram() =
  db.close()
```

Another remark: As far as possible, row data should be processed in place, instead of first copying the query result to a seq and then reiterating it again. If the result may contain a considerable amount of rows, taking a copy of the result will not compute well, and will ultimately result in OOM.

**Comment by ITwrx** [comment]
I read the main example several times, but never understood why you were handling prepareStatements() that way. I thought it was because sometimes you might want to execute _unprepared_ statements, which didn't seem relevant to me. I even had to read your last comment a few times, before what you were trying to tell me, sunk in. I thought, "why is he adding that unnecessary `prepared` check/condition to the example again?". 🤦🏻

You might consider adding a comment to the example(s) that mentions why you are doing that. In-context comments help a lot, and may save the user, and yourself, some time. 

The:
```Nim
db.openDatabase("ex.db", Schema)
defer: db.close()
```
might could use a comment explaining _why_ you are doing it that way too. I didn't catch that either. :)

Finally, it might be preferable to use the second `openDatabase()` proc with more than one schema in the example, since most programs will need more than one, especially since you will have to include all the schemas with the one allowed `openDatabase()` call.

I'll update once i reorganize my code. Just wanted to provide this UX/DX feedback.

**Comment by olliNiinivaara** [comment]
Your feedback is appreciated. Quality of documentation is the Achilles' heel of FOSS projects (the boring part, usually taken care with a slightly sarcastic "PRs welcome" comment :smiley:).

There's terminological confusion (my bad): a sqlite database always has exactly one schema. The openDatabase proc really just accepts a list of strings and calls ```exes``` with each of them (where each string usually contains multiple DDL (CREATE TABLE etc.) clauses, separated by commas). The idea is that when database schema is modified, new set of idempotent (that is why ignoring errors is supported)  clauses is added. Experience tells that running all DDL clauses at program start is a good sanity check.

So, until you need to do [schema migration](https://en.wikipedia.org/wiki/Schema_migration), one schema string should be enough.

...but still nothing prevents you from executing data definition clauses also after the db is open, like:
```nim
db.transaction:
  db.exes("ALTER TABLE Example DROP COLUMN big ; CREATE UNIQUE INDEX IF NOT EXISTS idx_string ON Example(string)")

```

**Comment by ITwrx** [comment]
Right, i was using the terminology of "schema(s)" in the same way the sqliteral proc was. Since both examples just show one table statement as the schema string constant passed to openDatabase(), i wasn't sure what the intended usage was in a real application. tiny_sqlite's openDatabase() doesn't require any schema passed at that time. Also, I hadn't scrolled down far enough in the sqliteral openDatabase() docs. I think i have it now.

I got my test working fully now, thanks to your help.
